### PR TITLE
Addon-a11y: use channel to highlight elements in preview

### PR DESCRIPTION
**Pull request by tooppaaa**
Issue:

The a11y addon is using a getIframe method to access the preview which won't work with storybook composition.

## What I did
 
- Refactor A11yPanel & HighlightToggle to use a context to move the logic out of these components
- Create A11yContext in charge of handling results, what's highlighted, some addon behavior
- Unit test
- Use channel to highlight element in the preview
- use additional stylesheet to highlight instead of playing directly with element's outline style

## How to test

- Highlight elements in a11y addon

### CODE REVIEW

File: addons/a11y/src/a11yHighlight.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+  const sheet = document.createElement('style');
+  sheet.setAttribute('id', id);
+  sheet.innerHTML = infos.elements
+    .map((target) => `${target}{ outline: 1px dotted ${infos.color}!important; }`)
+    .join(' ');
+  document.head.appendChild(sheet);
+};
+
+channel.on(EVENTS.HIGHLIGHT, highlight);
```
Comment (ndelangen) [inline addons/a11y/src/a11yHighlight.ts:32]: Does this need to be reset when `story_change` hanppens?

File: addons/a11y/src/redux-config.tsx [hunk 2]
```diff
@@ -1,44 +0,0 @@
-import { createStore } from 'redux';
```
Comment (ndelangen) [inline addons/a11y/src/redux-config.tsx:1]: Can we remove these dependencies now?

File: addons/a11y/src/components/A11YPanel.tsx [hunk 3]
```diff
@@ -49,228 +48,133 @@ const Centered = styled.span<{}>({
...
-          if (status === 'ran') {
-            this.setState({
-              status: 'ready',
-            });
-          }
-        }, 900);
+type Status = 'initial' | 'manual' | 'running' | 'error' | 'ran' | 'ready';
+
+export const A11YPanel: React.FC = () => {
+  const [status, setStatus] = React.useState<Status>('initial');
```
Comment (ndelangen) [inline addons/a11y/src/components/A11YPanel.tsx:54]: We can use `useAddonState` perhaps, to have it persist automatically by storing the state in the storybook global state?

File: addons/a11y/src/components/A11YPanel.tsx [hunk 4]
```diff
@@ -49,228 +48,133 @@ const Centered = styled.span<{}>({
...
+          <RotatingIcon inline icon="sync" /> Please wait while the accessibility scan is running
+          ...
+        </Centered>
+      )}
+      {(status === 'ready' || status === 'ran') && (
+        <>
+          <ScrollArea vertical horizontal>
+            <Tabs
+              key="tabs"
+              tabs={[
```
Comment (ndelangen) [inline addons/a11y/src/components/A11YPanel.tsx:113]: this creates a new list every render, we might want to extract into a `useMemo`?

File: addons/a11y/src/components/A11YPanel.tsx [hunk 5]
```diff
@@ -49,228 +48,133 @@ const Centered = styled.span<{}>({
...
-            {error}
-          </Centered>
-        );
-    }
-  }
-}
+          </ScrollArea>
+          <ActionBar
+            key="actionbar"
+            actionItems={[
```
Comment (ndelangen) [inline addons/a11y/src/components/A11YPanel.tsx:155]: extract into `useMemo` as well?

File: addons/a11y/src/a11yHighlight.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+interface HighlightInfo {
+  /** html selector of the element */
+  elements: string[];
+  color: string;
+}
+
+const channel = addons.getChannel();
+
+const highlight = (infos: HighlightInfo) => {
+  const id = 'a11yHighlight';
```
Comment (ndelangen) [inline addons/a11y/src/a11yHighlight.ts:18]: extract into addon `constants.ts`

File: addons/a11y/src/a11yHighlight.ts [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+  const sheetToBeRemoved = document.getElementById(id);
+  if (sheetToBeRemoved) {
+    sheetToBeRemoved.parentNode.removeChild(sheetToBeRemoved);
+  }
+
+  const sheet = document.createElement('style');
+  sheet.setAttribute('id', id);
+  sheet.innerHTML = infos.elements
+    .map((target) => `${target}{ outline: 1px dotted ${infos.color}!important; }`)
```
Comment (ndelangen) [inline addons/a11y/src/a11yHighlight.ts:27]: Ow nice, so the color in dependent on the priority?

Real meta, but how do we make sure those outlines are actually visible / accessible?

We should add a box-shadow perhaps of 1px background behind the dotted  outline?

File: addons/a11y/src/a11yHighlight.ts (hunk 1 above)
Comment (tooppaaa) [reply addons/a11y/src/a11yHighlight.ts:32]: Good catch. Done by A11yContext & unit tested

File: addons/a11y/src/redux-config.tsx (hunk 2 above)
Comment (tooppaaa) [reply addons/a11y/src/redux-config.tsx:1]: Yes

File: addons/a11y/src/a11yHighlight.ts (hunk 7 above)
Comment (tooppaaa) [reply addons/a11y/src/a11yHighlight.ts:27]: That's actually a good question. I'm not a big fan of the outline, I can't see the difference between red / green. Maybe we can have this is the a11y parameters ? (different PR), not sure how the theme impacts this

File: addons/a11y/src/components/A11YPanel.tsx (hunk 3 above)
Comment (tooppaaa) [reply addons/a11y/src/components/A11YPanel.tsx:54]: Yep, good idea

File: addons/a11y/src/a11yHighlight.ts (hunk 7 above)
Comment (ndelangen) [reply addons/a11y/src/a11yHighlight.ts:27]: Assume no theming for now, just static colors.

Comment (tooppaaa) [reply addons/a11y/src/a11yHighlight.ts]: What do you think of this ?
image

File: addons/a11y/src/a11yRunner.ts [hunk 8]
```diff
@@ -18,6 +18,7 @@ const getElement = () => {
 };
 
 const run = async (storyId: string) => {
+  console.log('Running', storyId);
```
Comment (shilman) [inline addons/a11y/src/a11yRunner.ts]: nope 😉

Comment (tooppaaa) [reply addons/a11y/src/a11yRunner.ts]: oh :)

File: addons/a11y/src/a11yHighlight.ts (hunk 7 above)
Comment (ndelangen) [reply addons/a11y/src/a11yHighlight.ts]: Made a tweak to this, WDYT?

Comment (tooppaaa) [reply addons/a11y/src/a11yHighlight.ts]: Good for me ;)

File: addons/a11y/package.json [hunk 9]
```diff
@@ -46,13 +46,10 @@
     "global": "^4.3.2",
     "lodash": "^4.17.15",
     "memoizerific": "^1.11.3",
-    "react-redux": "^7.0.2",
     "react-sizeme": "^2.5.2",
-    "redux": "^4.0.1",
```
Comment (ndelangen) [inline addons/a11y/package.json:51]: reducing the weight of the monorepo, 1 dependency at a time 💪

File: addons/a11y/src/a11yHighlight.ts [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+const channel = addons.getChannel();
+
+const higlighted = (color: string) => `
+  outline: 2px dashed ${color};
+  outline-offset: 2px;
+  box-shadow: 0 0 0 6px rgba(255,255,255,0.6);
+}
+`;
+
+const highlight = (infos: HighlightInfo) => {
```
Comment (ndelangen) [inline addons/a11y/src/a11yHighlight.ts:24]: Let's extract into a different module and re-sue it in a story so we can test this visually.

File: addons/a11y/src/a11yHighlight.ts [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+  color: string;
+}
+
+const channel = addons.getChannel();
+
+const higlighted = (color: string) => `
+  outline: 2px dashed ${color};
+  outline-offset: 2px;
+  box-shadow: 0 0 0 6px rgba(255,255,255,0.6);
+}
```
Comment (ndelangen) [inline addons/a11y/src/a11yHighlight.ts:21]: Let's extract into a different module and re-sue it in a story so we can test this visually.

Comment (tooppaaa) [reply addons/a11y/src/a11yHighlight.ts:22]: Done !

**Review by ndelangen — APPROVED** [review]

**Comment by CodeByAlex** [comment]
Great work on this! Excited to see it switched over to use some better practices

**Pull request merged**